### PR TITLE
CI scripts/pipeline clean up

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR is a clean up effort after the various changes made in the past 3 months. It reuses the new helper functions in the Jenkinsfile as much as possible, introduces a first script library to centralize how flaky tests are analysed for easier maintainability, moves the `test-compile` phase of the IT agent into its prepare stage so the test stage logs are only about testing, and makes sure the way we configure Maven in built/test scripts is consistent across all scripts.

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

**Comment by npepinpe** [comment]
Blocked by https://github.com/zeebe-io/zeebe/pull/6020 - specifically the first two commit belong to that branch.

**Comment by npepinpe** [comment]
TODO: inject fake flaky tests to make sure flaky test detection isn't broken.

File: Jenkinsfile [hunk 1]
```diff
@@ -248,20 +259,19 @@ pipeline {
...
                                 timeout(time: longTimeoutMinutes, unit: 'MINUTES') {
-                                    unstash name: "zeebe-build"
                                     runMavenContainerCommand('.ci/scripts/distribution/it-java.sh')
                                 }
                             }
 
                             post {
                                 always {
                                     junit testResults: "**/*/TEST*${SUREFIRE_REPORT_NAME_SUFFIX}*.xml", keepLongStdio: true
+                                    stash allowEmpty: true, name: itFlakyTestStashName, includes: '**/FlakyTests.txt'
```
Comment (npepinpe) [inline Jenkinsfile:269]: Moved here to avoid error if there are only flaky tests in the main agent, since you cannot test if a stash is present or not

**Review by korthout — APPROVED** [review]
Nice work on this! It LGTM, but please verify my only concern.

File: .ci/scripts/distribution/test-java8.sh [hunk 2]
```diff
@@ -1,29 +1,28 @@
...
+  -DtestMavenId=3
+  -Dsurefire.rerunFailingTestsCount=7
+)
 tmpfile=$(mktemp)
 
-mvn -o -B --fail-never -T$LIMITS_CPU -s ${MAVEN_SETTINGS_XML} verify -pl clients/java -DtestMavenId=3 -Dsurefire.rerunFailingTestsCount=7 | tee ${tmpfile}
+if [ ! -z "$SUREFIRE_FORK_COUNT" ]; then
+  MAVEN_PROPERTIES+=("-DforkCount=$SUREFIRE_FORK_COUNT")
+  # if we know the fork count, we can limit the max heap for each fork to ensure we're not OOM killed
+  JAVA_TOOL_OPTIONS="${JAVA_TOOL_OPTIONS} -XX:MaxRAMPercentage=$((100 / ($MAVEN_PARALLELISM * $SUREFIRE_FORK_COUNT)))"
```
[the remark this answers is not in the document]
Comment (npepinpe) [reply .ci/scripts/distribution/test-java8.sh:18]: Ah, you're right it should be exported! :see_no_evil:

[the remark this answers is not in the document]
Comment (npepinpe) [reply .ci/scripts/distribution/test-java8.sh]: Consequences of not exporting means that you could run out of memory (the famous error code 137), but it seems not to have affected us too much.

**Pull request merged**